enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,405 @@
+investigation is needed, but it's likely that we can not test this on our
+`e2e-metal` UPI based jobs because we rely on the network provided by
+packet.net between the cluster hosts.
+
+Testing of the BGP mode is more complicated.  It will require setting up a BGP
+network environment for the cluster nodes to peer with.  We don't have anything
+like this today, so it will take some work.  The upstream MetalLB project needs
+this, as well.  It currently lacks any automated testing of the BGP
+integration.
+
+### Upgrade / Downgrade Strategy
+
+The operator will include any required logic to handle upgrades or downgrades
+and changes between versions of MetalLB.
+
+MetalLB is currently configured via a `ConfigMap` that is not a stable API.  We
+will start by building an operator that provides a stable API for configuration
+and the `ConfigMap` will become an internal implementation detail fully owned
+by the operator.
+
+We must also mitigate these risks through engagement and contributions to the
+upstream community to help make sure that changes made to the software and its
+configuration interfaces can be managed through an upgrade or downgrade
+process.
+
+### Version Skew Strategy
+
+MetalLB has two major components: a single `controller` pod and a `speaker` pod
+that runs as a `DaemonSet`.
+
+In layer 2 mode, the `speaker` needs to run on every `Node`.
+
+In BGP mode, there's more flexibility. The behavior depends on the
+[ExternalTrafficPolicy](https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip)
+type on the service load balancers.
+
+* If the type is `cluster`, then `speaker` can run on any subset of Nodes that
+  you want peering with BGP routers. All of those Nodes will advertise that
+  they can be used to reach all of of the Service load balancer IPs.
+* If the type is `local`, then `speaker` must run on every node where an Endpoint
+  may exist locally. Each node will only advertise that load balancer IPs are
+  reachable that can map to a local Endpoint. Put another way, speaker must run
+  on every node where workloads behind a service load balancer with an
+  `ExternalTrafficPolicy` of `local` may run.
+
+The primary version skew concern would be when there are `speaker` instances
+running from different versions of MetalLB.  For example, in the layer 2 mode,
+the `speaker` implementation includes an algorithm for it to independently
+determine whether it should be the leader, or announcer, for a given `Service`.
+A change to this algorithm in a new version could cause more than one `speaker`
+to think it owns a `Service`.
+
+These risks to version skew must be mitigated through upstream community
+engagement and contributions, as well as informed management of upgrades in the
+MetalLB operator.
+
+## Implementation History
+
+* (May, 2020) - Technical due diligence and upstream engagement beginning
+
+## Drawbacks
+
+TBD
+
+## Alternatives
```

Comment:
I went down this path some months ago and let me share my findings, excuse me if I say something wrong because I don't have a good knowledge of Openshift yet.
I've come to the conclusion that the solution for bare metal should be based in a cloud-controller-manager, implementing the cloudprovider.Interface https://github.com/kubernetes/cloud-provider/blob/master/plugins.go and reusing the common logic there. Then, you can use whatever implementation for the load balancer, a good example of this approach is the `Packet` cloud controller manager, that uses MetalLB under the hood https://github.com/packethost/packet-ccm

It's a very neat approach IMHO with the benefit that you have a clear interface already defined upstream

Reply:
@aojea great feedback, thank you!  I will explore this in more detail.

Reply:
> I've come to the conclusion that the solution for bare metal should be based in a cloud-controller-manager, implementing the cloudprovider.Interface

But you can only have one cloud provider, can't you? The doc says they ideally want something that will work with OpenStack and VMware too, where presumably you'd want to keep using the native (load-balancer-less) cloud provider.

Reply:
ok, my fault, I should've have explained myself better, because I'm considering that we have a control-plane piece (cloud-controller-manager) and a data-plane, envoy, metallb uses bgp or keepalived like load balancing, ..ovn , 
I think makes sense to have a different cloud-controller-manager per environment, i.e aws, gcp, azure, openstck, vsphere, ..., actually this is how it is working upstream.
However, the dataplane mechanism may be the same for all ... if the load balancer is going to be as simple as metallb it makes sense to use OVN, just bear in mind that if you expect demand for real load balancing features (SSL/TLS, grpc, affinity, health checking, ... ) then envoy is the king there and we start to move to the service mesh area ... 
It really depends on what you mean by load balancer: keepalived or envoy, for me load balancer is the later

Reply:
I've added the CCM integration as an alternative integration path, but not specified what the conclusion is.  I assume that would come in a future update to this enhancement which proposes more detailed integration plans once further investigation is complete.

Reply:
(I realize this is marked resolved, but I wanted to document the discussions we've had offline)

One possible issue with using a CCM is that in metal3 we don't have a cloud API to talk to. Using a CCM would likely require work with the metal3 community to figure out how/if that would work.